File: src/plugins/platforms/kook/normalizer.py
import json
import re
from typing import List

from khl import Bot, Message, MessageTypes

from src.core.entities.message import Message
from src.core.entities.transport.actor import ActorRef
from src.core.entities.transport.attachments import Attachment
from src.core.entities.transport.envelope import MessageEnvelope
from src.core.entities.transport.transport import TransportRef
from src.core.values import (
    ActorRefType,
    AttachmentKind,
    MessageEnvelopeDirectionType,
    MessageEnvelopeTransportFlowType,
    now_ms,
)

__all__ = ["KookNormalizer"]


class KookNormalizer:
    """Convert KOOK messages into the application's ``MessageEnvelope`` model."""

    def __init__(self, bot: Bot) -> None:
        """Initialize the KOOK message normalizer.

        Args:
            bot: KOOK bot used to identify the current bot account.
        """
        self._bot = bot

    async def normalize(self, msg: Message) -> MessageEnvelope | None:
        """Normalize an incoming KOOK message into a ``MessageEnvelope``.

        Messages sent by bots, messages that do not mention the current bot, and
        malformed card messages are ignored.

        Args:
            msg: Incoming KOOK message to normalize.

        Returns:
            The normalized envelope, or ``None`` if the message should be ignored.
            ``conversation_id`` is left unset here (platform-agnostic mapping); it
            is resolved by the inbound use case against a ``Conversation``.
        """
        # Skip the robot itself and messages that do not mention the bot.
        bot_user = await self._bot.client.fetch_me()
        # if getattr(msg.author, "bot", False) or msg.author_id == bot_user.id:
        #     return None
        if bot_user.id not in (msg.extra.get("mention") or []):
            return None

        text = msg.content or ""
        attachments: List[Attachment] = []

        if msg.type == MessageTypes.KMD:
            text = (msg.extra.get("kmarkdown") or {}).get("raw_content") or text
        elif msg.type == MessageTypes.CARD:
            text_parts = []
            try:
                cards = json.loads(text)
            except (TypeError, json.JSONDecodeError):
                return None
            for card in cards:
                for module in card.get("modules", []):
                    module_type = module.get("type")
                    if module_type == "section":
                        content = (module.get("text") or {}).get("content")
                        if content:
                            text_parts.append(content)
                        accessory = module.get("accessory") or {}
                        if accessory.get("type") == "image" and accessory.get("src"):
                            attachments.append(_image_attachment(accessory))
                    elif module_type in {"container", "image-group", "context"}:
                        for element in module.get("elements", []):
                            src = element.get("src")
                            if src:
                                attachments.append(_media_attachment(element, src))
                    elif module_type in {"file", "audio", "video"} and module.get("src"):
                        attachments.append(_media_attachment(module, module["src"]))
            text = "\n".join(text_parts)
        text = re.sub(rf"\(met\){re.escape(bot_user.id)}\(met\)", "", text)
        text = re.sub(rf"<@!?{re.escape(bot_user.id)}>", "", text).strip()

        now = now_ms()
        return MessageEnvelope(
            conversation_id=None,  # resolved by the inbound use case
            messages=Message(
                content=text or None,
                attachments=attachments, 
                reply_to_id=None,
                sent_at=now),
            sender=ActorRef(
                external_id=msg.author.id,
                actor_type=ActorRefType.PEOPLE,
                display_name=msg.author.nickname,
                avatar=msg.author.vip_avatar
            ),
            # recipient=,
            transport=TransportRef(
                external_event_id=msg.id,
                external_message_id=msg.id,
            ),
            direction=MessageEnvelopeDirectionType.P2B,
            transport_flow=MessageEnvelopeTransportFlowType.INBOUND,
            received_at=now,
            idempotency_key=msg.id,
        )


def _image_attachment(accessory: dict) -> Attachment:
    """Build an image Attachment from a section accessory."""
    return Attachment(
        f_type=AttachmentKind.IMAGE,
        name=accessory.get("alt"),
        source_url=accessory.get("src"),
    )


def _media_attachment(node: dict, src: str) -> Attachment:
    """Build an Attachment from a media element/file/audio/video node."""
    node_type = node.get("type")
    if node_type == "image":
        kind = AttachmentKind.IMAGE
    elif node_type == "audio":
        kind = AttachmentKind.AUDIO
    elif node_type == "video":
        kind = AttachmentKind.VIDEO
    elif node_type in {"file", "image-group", "container", "context"}:
        kind = AttachmentKind.FILE
    else:
        kind = AttachmentKind.FILE
    return Attachment(
        f_type=kind,
        name=node.get("title") or node.get("name"),
        mime_type=node.get("mime_type"),
        source_url=src,
    )

File: packages/platform-kook/src/normalizer.ts
import {
  ActorRef,
  ActorRefType,
  Attachment,
  AttachmentKind,
  BotPlatform,
  ConversationAddress,
  Message,
  MessageEnvelope,
  MessageEnvelopeDirection,
  MessageRole,
  TransportFlow,
  TransportRef,
  nowMs,
  unixMillis,
  type NormalizedIncoming,
} from "@kookbot/domain";

export enum KookMessageType {
  TEXT = 1,
  IMAGE = 2,
  VIDEO = 3,
  FILE = 4,
  AUDIO = 8,
  KMARKDOWN = 9,
  CARD = 10,
}

export interface KookUser {
  readonly id: string;
  readonly username?: string;
  readonly nickname?: string;
  readonly avatar?: string;
  readonly vip_avatar?: string;
  readonly bot?: boolean;
}

export interface KookMessageEvent {
  readonly channel_type: "GROUP" | "PERSON" | "BROADCAST";
  readonly type: number;
  readonly target_id: string;
  readonly author_id: string;
  readonly content: string;
  readonly msg_id: string;
  readonly msg_timestamp: number;
  readonly nonce?: string;
  readonly extra: {
    readonly [key: string]: unknown;
    readonly type: number;
    readonly guild_id?: string;
    readonly mention?: readonly string[];
    readonly author?: KookUser;
    readonly attachments?:
      | { readonly type: string; readonly url: string; readonly name?: string; readonly size?: number }
      | readonly { readonly type: string; readonly url: string; readonly name?: string; readonly size?: number }[];
    readonly quote?: { id: string };
    readonly kmarkdown?: { raw_content?: string };
  };
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&");
}

function attachmentKind(type: string): AttachmentKind {
  if (type === "image") return AttachmentKind.IMAGE;
  if (type === "audio") return AttachmentKind.AUDIO;
  if (type === "video") return AttachmentKind.VIDEO;
  return AttachmentKind.FILE;
}

function messageAttachmentKind(type: number): AttachmentKind {
  if (type === KookMessageType.IMAGE) return AttachmentKind.IMAGE;
  if (type === KookMessageType.AUDIO) return AttachmentKind.AUDIO;
  if (type === KookMessageType.VIDEO) return AttachmentKind.VIDEO;
  return AttachmentKind.FILE;
}

function eventAttachments(value: KookMessageEvent["extra"]["attachments"]): Attachment[] {
  const items = value === undefined ? [] : Array.isArray(value) ? value : [value];
  return items.map(
    (item) =>
      new Attachment({
        kind: attachmentKind(item.type),
        sourceUrl: item.url,
        name: item.name,
        ...(item.size === undefined ? {} : { size: item.size }),
      }),
  );
}

function extractCard(content: string): { text: string; attachments: Attachment[] } | undefined {
  let cards: unknown;
  try {
    cards = JSON.parse(content);
  } catch {
    return undefined;
  }
  if (!Array.isArray(cards)) return undefined;
  const texts: string[] = [];
  const attachments: Attachment[] = [];
  for (const card of cards) {
    if (!card || typeof card !== "object" || !("modules" in card) || !Array.isArray(card.modules)) continue;
    for (const rawModule of card.modules) {
      if (!rawModule || typeof rawModule !== "object") continue;
      const module = rawModule as Record<string, unknown>;
      if (module.type === "section") {
        const text = module.text as { content?: unknown } | undefined;
        if (typeof text?.content === "string") texts.push(text.content);
        const accessory = module.accessory as Record<string, unknown> | undefined;
        if (accessory?.type === "image" && typeof accessory.src === "string") {
          attachments.push(
            new Attachment({
              kind: AttachmentKind.IMAGE,
              sourceUrl: accessory.src,
              ...(typeof accessory.alt === "string" ? { name: accessory.alt } : {}),
            }),
          );
        }
      }
      const elements = Array.isArray(module.elements) ? module.elements : [];
      for (const rawElement of elements) {
        if (!rawElement || typeof rawElement !== "object") continue;
        const element = rawElement as Record<string, unknown>;
        if (typeof element.src !== "string") continue;
        attachments.push(
          new Attachment({
            kind: attachmentKind(typeof element.type === "string" ? element.type : "file"),
            sourceUrl: element.src,
            ...(typeof element.title === "string" ? { name: element.title } : {}),
          }),
        );
      }
      if (["file", "audio", "video"].includes(String(module.type)) && typeof module.src === "string") {
        attachments.push(
          new Attachment({
            kind: attachmentKind(String(module.type)),
            sourceUrl: module.src,
            ...(typeof module.title === "string" ? { name: module.title } : {}),
          }),
        );
      }
    }
  }
  return { text: texts.join("\n"), attachments };
}

function occurredAt(timestamp: number): ReturnType<typeof unixMillis> {
  const milliseconds = timestamp < 10_000_000_000 ? timestamp * 1000 : timestamp;
  return unixMillis(Math.max(0, Math.trunc(milliseconds)));
}

export class KookMessageNormalizer {
  normalize(event: KookMessageEvent, botUserId: string): NormalizedIncoming | undefined {
    const author = event.extra.author;
    if (!author?.id || author.bot === true || author.id === botUserId) return undefined;
    const direct = event.channel_type === "PERSON";
    if (!direct && !event.extra.mention?.includes(botUserId)) return undefined;

    let text = event.content ?? "";
    let attachments = eventAttachments(event.extra.attachments);
    if (
      [KookMessageType.IMAGE, KookMessageType.VIDEO, KookMessageType.FILE, KookMessageType.AUDIO].includes(
        event.type,
      ) &&
      /^https?:\/\//u.test(text)
    ) {
      if (!attachments.some((attachment) => attachment.sourceUrl === text)) {
        attachments.push(new Attachment({ kind: messageAttachmentKind(event.type), sourceUrl: text }));
      }
      text = "";
    }
    if (event.type === KookMessageType.KMARKDOWN) text = event.extra.kmarkdown?.raw_content ?? text;
    if (event.type === KookMessageType.CARD) {
      const card = extractCard(text);
      if (!card) return undefined;
      text = card.text;
      attachments = [...attachments, ...card.attachments];
    }
    const escapedBotId = escapeRegExp(botUserId);
    text = text
      .replace(new RegExp(`\\(met\\)${escapedBotId}\\(met\\)`, "gu"), "")
      .replace(new RegExp(`<@!?${escapedBotId}>`, "gu"), "")
      .trim();

    const address = direct
      ? new ConversationAddress({ platform: BotPlatform.KOOK, externalId: author.id })
      : new ConversationAddress({
          platform: BotPlatform.KOOK,
          roomId: event.target_id,
          ...(event.extra.guild_id === undefined ? {} : { spaceId: event.extra.guild_id }),
        });
    return {
      address,
      envelope: new MessageEnvelope({
        message: new Message({ role: MessageRole.USER, ...(text ? { content: text } : {}), attachments }),
        sender: new ActorRef({
          actorType: ActorRefType.PEOPLE,
          externalId: author.id,
          displayName: author.nickname || author.username,
          avatar: author.vip_avatar || author.avatar,
        }),
        transport: new TransportRef({
          externalEventId: event.nonce || event.msg_id,
          externalMessageId: event.msg_id,
          ...(event.extra.quote?.id === undefined ? {} : { externalReplyToMessageId: event.extra.quote.id }),
        }),
        direction: MessageEnvelopeDirection.P2B,
        transportFlow: TransportFlow.INBOUND,
        occurredAt: occurredAt(event.msg_timestamp),
        receivedAt: nowMs(),
        idempotencyKey: `kook:${event.msg_id}`,
      }),
    };
  }
}
